pkg/target/manager.go

```diff
@@ -0,0 +1,399 @@
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		var (
+			project *gardencorev1beta1.Project
+			seed    *gardencorev1beta1.Seed
+		)
+
+		if t.Project != "" {
+			project, err = m.resolveProjectName(ctx, gardenClient, t.Project)
+			if err != nil {
+				return fmt.Errorf("failed to validate project: %w", err)
+			}
+		} else {
+			seed, err = m.resolveSeedName(ctx, gardenClient, t.Seed)
+			if err != nil {
+				return fmt.Errorf("failed to fetch kubeconfig for seed cluster: %w", err)
+			}
+		}
+
+		t.Shoot = shootName
+
+		if err := m.updateShootKubeconfig(ctx, gardenClient, project, seed, t); err != nil {
+			return fmt.Errorf("failed to fetch kubeconfig for shoot cluster: %w", err)
+		}
+
+		return nil
+	})
+}
+
+// updateShootKubeconfig must be called with *either* project or seed, but never both
+// and never neither.
+func (m *managerImpl) updateShootKubeconfig(
+	ctx context.Context,
+	gardenClient client.Client,
+	project *gardencorev1beta1.Project,
+	seed *gardencorev1beta1.Seed,
+	t Target,
+) error {
+	shoot := &gardencorev1beta1.Shoot{}
+
+	// If a shoot is targeted via a project, we fetch the project and find
+	// the shoot by listing all shoots in the project's spec.namespace.
+	// If the target uses a seed, _all_ shoots in the garden are filtered
+	// for shoots with matching seed and name. It's an error if no or multiple
+	// matching shoots are found.
+
+	if project != nil {
+		// fetch shoot from project namespace
+		// TODO: a project's spec.namespace can be nil, what to do about those cases?
+		key := types.NamespacedName{Name: t.ShootName(), Namespace: *project.Spec.Namespace}
+
+		if err := gardenClient.Get(ctx, key, shoot); err != nil {
+			return fmt.Errorf("invalid shoot %q: %w", key.Name, err)
+		}
+	} else if seed != nil {
+		// list all shoots, filter by spec.seedName
+		shootList := gardencorev1beta1.ShootList{}
+		if err := gardenClient.List(ctx, &shootList, &client.ListOptions{}); err != nil {
+			return fmt.Errorf("failed to list shoot clusters: %w", err)
+		}
+
+		// filter found shoots; if multiple shoots have the same name, but are in different
+		// projects, we have to warn the user about their ambiguous targeting
+		matchingShoots := []*gardencorev1beta1.Shoot{}
+		for i, s := range shootList.Items {
+			if s.Name == t.ShootName() && s.Spec.SeedName != nil && *s.Spec.SeedName == seed.Name {
+				matchingShoots = append(matchingShoots, &shootList.Items[i])
+			}
+		}
+
+		if len(matchingShoots) == 0 {
+			return fmt.Errorf("invalid shoot %q: not found on seed %q", t.ShootName(), seed.Name)
+		}
+
+		if len(matchingShoots) > 1 {
+			return fmt.Errorf("there are multiple shoots named %q on this garden, please target via project to make your choice unambiguous", t.ShootName())
+		}
+
+		shoot = matchingShoots[0]
+	} else {
+		return errors.New("neither project nor seed were provided to filter the shoot by")
```

Comment:
well basically, similar as when a seed is targeted, we can go over all the shoots to find it

Reply:
Sure, I just replicated the v1 behaviour. Not sure if we want to deviate already from the old behaviour.